### ISSUE TITLE
FFmpeg has a memory leak in OSMC for raspberry pi 4

### ISSUE THREAD

**Issue by Minauras** [osmc/osmc#679]
### Bug description
When playing videos with Kodi in either OSMC or Xbian on a Raspberry Pi CM4, a small amount of allocated heap memory is not freed when the video ends. These leaks build up and cause Kodi to crash after playing several videos.

Changing the allocated heap size with the cma option in either `/boot/config.txt` or `/boot/cmdline.txt` does not solve the problem, it only changes when it happens.

## Possible Fix
This behavior is not present on LibreELEC. Team Kodi member popcornmix on the [corresponding github issue](https://github.com/xbmc/xbmc/issues/20908) suggests OSMC's ffmpeg version is not up to date. Updating ffmpeg may fix the issue.

### To Reproduce
Setup: Raspberry Pi CM4 on TOFU carrier board, with either OSMC or Xbian.

Either:
- Run a video playlist on loop and monitor the RAM or heap. Each time a video starts, the available memory drops. When the available memory reaches a critical state, Kodi crashes and restarts.
- Run a video and stop it, then do it again on loop. Monitor the RAM or heap. Each time the video starts, the available memory drops. When the available memory reaches a critical state, Kodi crashes and restarts.

### Debug log
[Here is a full log from a run with this bug.](https://gist.github.com/Minauras/bf71b945abc4acd3b299fbc2d68c9591)
The error is the following:
```
2022-01-21 14:34:33.951 T:1510     INFO <general>: ffmpeg[(nil)X]: Failed to alloc 12533760 from dma-heap(fd=176): 12 (Cannot allocate memory)
2022-01-21 14:34:33.951 T:1510    ERROR <general>: ffmpeg[(nil)X]: [hevc] v4l2_request_hevc_start_frame: Failed to get dst buffer
```

## Additional context
I am playing 4k24fps videos, I have not tried with other video formats.

I first reported this issue on the [Kodi forum](https://forum.kodi.tv/showthread.php?tid=366631&pid=3082917#pid3082917) and the [OSMC forum](https://discourse.osmc.tv/t/playling-playlist-dies-after-about-7-minutes-possible-memory-leak/92347).

I also raised this issue on the [XBMC github tracker](https://github.com/xbmc/xbmc/issues/20908) and on the [rpi-ffmpeg tracker](https://github.com/jc-kynesim/rpi-ffmpeg/issues/33).

### Environment
 - Operating system version/name: Either OSMC 2021.12-1 or Xbian 1.0 2021-11-22
 - Kodi version: 19.3.0
 - Hardware: Raspberry Pi CM4 on TOFU carrier board

**Comment by popcornmix** [comment]
@jc-kynesim has indicated a memory leak was fixed on 2021-10-07.
The commit adds a function `v4l2_req_hevc_alloc_frame`.
That commit is not present in https://github.com/osmc/osmc/blob/master/package/mediacenter-osmc/patches/rbp-000-add-pi-support.patch
It is present in ffmpeg patch from https://github.com/popcornmix/xbmc/tree/gbm_matrix

**Comment by samnazarko** [comment]
We can get this in and re-sync with Kodi v19.4.

**Comment by jc-kynesim** [comment]
Just to be clear - that was just one of a sequence of memory leak fixes that occurred round about then - you need to get the whole lot there's no simple cherry-pick.

**Comment by Minauras** [comment]
Hi and thanks for the new OSMC version!

I tried it, but unfortunately, the problem is still present in OSMC version 2022.03-1.

Is there any hope to see this issue fixed in the near future?

Bests, Nara :)

**Comment by samnazarko** [comment]
It will be solved with the next upstream merge.

**Comment by Minauras** [comment]
Hi, I'm sorry to reopen this issue again.

After stumbling upon your [testing post](https://discourse.osmc.tv/t/testing-debian-11-bullseye/93054) on the OSMC forum, I decided to check on the newer dev versions, and unfortunately,  the memory leak is still there.

I tried with a fresh install of OSMC_TGT_rbp4_202205.27.img.gz, and then by updating this install using your instructions from the post:
> - `echo 'deb http://apt.osmc.tv buster-devel main' | sudo tee /etc/apt/sources.list.d/osmc-devel.list`
> - Check for updates via the My OSMC interface by going to My OSMC → Updates → Manual Controls → Check for Updates now. Install the updates when prompted to do so.

In both cases, the memory leak happened and crashed Kodi.

I understand you're probably focusing on a lot of other issues that have higher priority right now, I just wanted to check if this issue was still on your radar.

Thanks a lot!
Nara
